nums/core/array/application.py

```diff
@@ -1206,3 +1205,89 @@ def column_stack(self, tup, axis_block_size=None):
             else:
                 arrays.append(self.atleast_2d(arr))
         return self.concatenate(arrays, 1, axis_block_size=axis_block_size)
+
+    def sort(self, arr, kind):
+        # If the BlockArray to be sorted is small enough, default to serial sort
+        if arr.blocks.size == 1:
+            oid = self.cm.sort(
+                arr.flattened_oids()[0],
+                kind,
+                syskwargs={
+                    "grid_entry": (0,),
+                    "grid_shape": (1,),
+                    "options": {"num_returns": 1},
+                },
+            )
+            return BlockArray.from_oid(oid, arr.block_shape, arr.dtype, self.cm)
+
+        num_blocks = arr.grid.grid_shape[0]
+        pivot_oids = []
+
+        # Sample n - 1 pivots to sort buckets around
+        for i, grid_entry in enumerate(arr.grid.get_entry_iterator()):
+            if i == 0:
+                continue
+            oid = self.cm.sample_pivots(
+                arr.blocks[i].oid,
+                syskwargs={
+                    "grid_entry": grid_entry,
+                    "grid_shape": arr.grid.grid_shape,
+                    "options": {"num_returns": 1},
+                },
+            )
+            pivot_oids.append(BlockArray.from_oid(oid, (1,), arr.dtype, self.cm))
+
+        ba_pivots = self.concatenate(
+            pivot_oids, axis=0, axis_block_size=len(pivot_oids)
+        )
+
+        # Assume that the number of pivots is small enough to fit within a single BlockArray
+
+        sorted_pivots_oid = self.cm.sort(
+            ba_pivots.blocks[0].oid,
+            kind,
+            syskwargs={
+                "grid_entry": (0,),
+                "grid_shape": (1,),
+                "options": {"num_returns": 1},
+            },
+        )
+
+        mapped_sorted = np.empty([num_blocks, num_blocks], dtype=object)
+        reduce_sorted = []
+
+        for i, grid_entry in enumerate(arr.grid.get_entry_iterator()):
+            mapped_sorted[i] = self.cm.map_sort(
+                arr.blocks[i].oid,
+                sorted_pivots_oid,
+                kind,
+                syskwargs={
+                    "grid_entry": grid_entry,
+                    "grid_shape": arr.grid.grid_shape,
+                    "options": {"num_returns": num_blocks},
+                },
+            )
+
+        oids = []
+
+        for j, grid_entry in enumerate(arr.grid.get_entry_iterator()):
+            oids.append(
+                self.cm.reduce_sort(
+                    *mapped_sorted[:, j],
+                    kind=kind,
+                    syskwargs={
+                        "grid_entry": grid_entry,
+                        "grid_shape": arr.grid.grid_shape,
+                        "options": {"num_returns": 1},
+                    },
+                )
+            )
+
+        for oid in oids:
+            shape = self.cm.get(oid).shape
```

Comment:
Here, you can call something like `shape_dtypes.append(self.cm.shape_dtype(oid, syskwargs={"grid_entry": grid_entry, "grid_shape": arr.grid.grid_shape}))` (use the same for loop as above). After the loop, call `self.cm.get(shape_dtypes)` to get a list of `(shape, dtype)` tuples. You can just use `zip` to transpose this into two lists of shapes and dtypes. You can then use the shapes in a final loop used to construct the individual block arrays.